File: Leetcode 3101-3200/3139/3139.java
// Leetcode 3139: Minimum Cost to Equalize Array
// https://leetcode.com/problems/minimum-cost-to-equalize-array/
// Solved on 12th of December, 2025
class Solution {
    /**
     * Calculates the minimum cost to equalize all elements in an array.
     * @param nums The input array of integers.
     * @param cost1 The cost to increment a single element by 1.
     * @param cost2 The cost to increment two distinct elements by 1.
     * @return The minimum cost to make all elements equal, modulo 1,000,000,007.
     */
    public int minCostToEqualizeArray(int[] nums, int cost1, int cost2) {
        long minVal = Integer.MAX_VALUE;
        long maxVal = Integer.MIN_VALUE;
        long totalSum = 0;
        int n = nums.length;

        for (int num : nums) {
            minVal = Math.min(minVal, num);
            maxVal = Math.max(maxVal, num);
            totalSum += num;
        }

        long totalGap = maxVal * n - totalSum;

        if (cost1 * 2 <= cost2 || n <= 2) {
            return (int) ((totalGap * cost1) % 1_000_000_007);
        }

        long minCost = Long.MAX_VALUE;

        for (long target = maxVal; target <= maxVal + 2; target++) {
            minCost = Math.min(minCost, calculateCost(target, n, totalSum, minVal, cost1, cost2));
        }

        long limit = (totalSum - 2 * minVal) / (n - 2);
        
        if (limit >= maxVal) {
            for (long target = limit - 1; target <= limit + 2; target++) {
                if (target >= maxVal) {
                    minCost = Math.min(minCost, calculateCost(target, n, totalSum, minVal, cost1, cost2));
                }
            }
        }

        return (int) (minCost % 1_000_000_007);
    }

    private long calculateCost(long target, int n, long totalSum, long minVal, int cost1, int cost2) {
        long totalGap = target * n - totalSum;
        long maxGap = target - minVal;
        long otherGaps = totalGap - maxGap;

        if (maxGap > otherGaps) {
            return otherGaps * cost2 + (maxGap - otherGaps) * cost1;
        } else {
            long pairs = totalGap / 2;
            long cost = pairs * cost2;
            if (totalGap % 2 != 0) {
                cost += cost1;
            }
            return cost;
        }
    }
}

File: Leetcode 3101-3200/3139/3139.py
# Leetcode 3139: Minimum Cost to Equalize Array
# https://leetcode.com/problems/minimum-cost-to-equalize-array/
# Solved on 12th of December, 2025
class Solution:
    def minCostToEqualizeArray(self, nums: list[int], cost1: int, cost2: int) -> int:
        """
        Calculates the minimum cost to make all elements in the array equal.
        :param nums: A list of integers representing the array.
        :param cost1: The cost to increment a single element by 1.
        :param cost2: The cost to increment two different elements by 1.
        :return: The minimum cost to equalize the array, modulo 10^9 + 7.
        """
        n = len(nums)
        if n == 1:
            return 0

        minVal = min(nums)
        maxVal = max(nums)
        totalSum = sum(nums)
        mod = 10**9 + 7

        if cost1 * 2 <= cost2:
            target = maxVal
            ops = target * n - totalSum
            return (ops * cost1) % mod

        def getCost(target):
            totalOps = target * n - totalSum
            maxDiff = target - minVal

            if maxDiff > totalOps - maxDiff:
                pairedOps = totalOps - maxDiff
                singleOps = maxDiff - pairedOps
                return pairedOps * cost2 + singleOps * cost1
            else:
                pairs = totalOps // 2
                rem = totalOps % 2
                return pairs * cost2 + rem * cost1

        res = float('inf')

        for target in range(maxVal, maxVal + 3):
            res = min(res, getCost(target))

        if n > 2:
            numerator = totalSum - 2 * minVal
            denominator = n - 2
            minTarget = (numerator + denominator - 1) // denominator
            base = max(maxVal, minTarget)

            for target in range(base - 2, base + 3):
                if target >= maxVal:
                    res = min(res, getCost(target))

        return res % mod